frontend/src/components/smart/WeaponGrid.vue

```diff
@@ -167,12 +191,42 @@ export default Vue.extend({
     saveFilters() {
       localStorage.setItem('weapon-starfilter', this.starFilter);
       localStorage.setItem('weapon-elementfilter', this.elementFilter);
+    },
+
+    toggleFavorite(e: Event, weaponId: number) {
+      e.preventDefault();
+      if(this.favorites[weaponId]) {
+        this.$delete(this.favorites, weaponId);
+      } else {
+        this.$set(this.favorites, weaponId, true);
+      }
+
+      localStorage.setItem('favorites', this.getFavoritesString(this.favorites));
+    },
+
+    getFavoritesString(favorites: Record<number, boolean>): string {
+      return JSON.stringify(favorites);
+    },
+
+    getFavoritesMap(favorites: string): Record<number, boolean> {
+      if(!favorites) {
+        return {};
+      }
+
+      const favoritesMap: Record<number, boolean> = {};
+      favorites.split(',').forEach(x => favoritesMap[+x] = true);
+      return favoritesMap;
+    },
+
+    isFavorite(weaponId: number): boolean {
+      return this.favorites[weaponId];
     }
   },
 
   mounted() {
     this.starFilter = localStorage.getItem('weapon-starfilter') || '';
     this.elementFilter = localStorage.getItem('weapon-elementfilter') || '';
+    this.favorites = JSON.parse(localStorage.getItem('favorites') as string);
```

Comment:
For next time, actually handle the null case instead of casting it away like this. This breaks when the item doesn't exist in localStorage.

Reply:
You're right, this caught me off guard, not so obvious it would fail, but makes sense after I looked into it, thanks.